File: frontend/src/api/grammar.ts
import apiClient from './client';

export interface GrammarReplacement {
  value: string;
}

export interface GrammarMatch {
  message: string;
  short_message: string;
  offset: number;
  length: number;
  rule_id: string;
  rule_issue_type: string;
  replacements: GrammarReplacement[];
}

export interface GrammarCheckResponse {
  matches: GrammarMatch[];
  text: string;
}

export const grammarApi = {
  async checkGrammar(
    projectId: number,
    text: string,
    language = 'en-US',
    documentId?: number,
    sectionId?: number,
  ): Promise<GrammarCheckResponse> {
    const { data } = await apiClient.post(`/projects/${projectId}/grammar/check`, {
      text,
      language,
      document_id: documentId,
      section_id: sectionId,
    });
    return data;
  },
};

File: backend/app/schemas/grammar.py
from pydantic import BaseModel
from typing import Optional


class GrammarCheckRequest(BaseModel):
    text: str
    language: str = "en-US"
    document_id: int | None = None
    section_id: int | None = None


class GrammarMatchReplacement(BaseModel):
    value: str


class GrammarMatch(BaseModel):
    message: str
    short_message: str
    offset: int
    length: int
    rule_id: str
    rule_issue_type: str
    replacements: list[GrammarMatchReplacement] = []


class GrammarCheckResponse(BaseModel):
    matches: list[GrammarMatch]
    text: str
